File: crossfirmarizer_client_python/src/crossfirmarizer/serial_packet.py
import struct

def cobs_encode(data: bytes) -> bytes:
    """Encodes a byte string using Consistent Overhead Byte Stuffing (COBS)."""
    encoded = bytearray()
    encoded.append(0)  # Header byte placeholder
    code_idx = 0
    code = 1
    
    for b in data:
        if b == 0:
            encoded[code_idx] = code
            code = 1
            code_idx = len(encoded)
            encoded.append(0)
        else:
            encoded.append(b)
            code += 1
            if code == 255:
                encoded[code_idx] = code
                code = 1
                code_idx = len(encoded)
                encoded.append(0)
                
    encoded[code_idx] = code
    return bytes(encoded)

def cobs_decode(encoded: bytes) -> bytes:
    """Decodes a COBS-encoded byte string."""
    decoded = bytearray()
    idx = 0
    
    while idx < len(encoded):
        code = encoded[idx]
        if code == 0:
            raise ValueError("Invalid COBS encoding: code byte cannot be zero.")
        
        idx += 1
        for i in range(code - 1):
            if idx >= len(encoded):
                raise ValueError("Invalid COBS encoding: code byte exceeds remaining data.")
            decoded.append(encoded[idx])
            idx += 1
        
        if code < 255 and idx < len(encoded):
            decoded.append(0)
    
    return bytes(decoded)

def create_packet(cmd: int, payload: bytes) -> bytes:
    """Builds a COBS-encoded serial packet with a 0x00 delimiter."""
    length = len(payload)
    
    # Calculate XOR checksum
    checksum = cmd ^ length
    for byte in payload:
        checksum ^= byte
        
    # Pack the unencoded data: CMD + LEN + PAYLOAD + CHK
    fmt = f'<BB{length}sB'
    raw_data = struct.pack(fmt, cmd, length, payload, checksum)
    
    # Encode with COBS and append the frame delimiter
    encoded_data = cobs_encode(raw_data)
    return encoded_data + b'\x00'

def parse_packet(encoded_packet: bytes):
    """Parses a COBS-encoded packet and returns the command, payload, and checksum."""
    if not encoded_packet.endswith(b'\x00'):
        raise ValueError("Packet must end with a 0x00 delimiter.")
    
    # Remove the trailing 0x00 and decode
    encoded_data = encoded_packet[:-1]
    raw_data = cobs_decode(encoded_data)
    
    if len(raw_data) < 3:
        raise ValueError("Decoded packet is too short to be valid.")
    
    cmd = raw_data[0]
    length = raw_data[1]
    
    if len(raw_data) != 2 + length + 1:
        raise ValueError("Decoded packet length does not match expected length.")
    
    payload = raw_data[2:2+length]
    checksum = raw_data[-1]
    
    # Verify checksum
    calculated_checksum = cmd ^ length
    for byte in payload:
        calculated_checksum ^= byte
        
    if checksum != calculated_checksum:
        raise ValueError("Checksum mismatch.")
    
    return cmd, payload, checksum

# Example Usage:
if __name__ == "__main__":
    command_id = 0x01
    
    # Notice we include 0x00 in the payload to prove COBS removes it
    data_to_send = bytes([0xFF, 0x00, 0x1A]) 
    
    packet = create_packet(command_id, data_to_send)
    print(f"Encoded Packet: {packet.hex().upper()}")

File: crossfirmarizer/Core/Src/serial_packet.c
#include "serial_packet.h"

// Decodes COBS data. Returns the length of the decoded data, or 0 on error.
static size_t cobs_decode(const uint8_t *input, size_t length, uint8_t *output)
{
    size_t read_index = 0;
    size_t write_index = 0;
    uint8_t code;
    uint8_t i;

    while (read_index < length)
    {
        code = input[read_index];
        if (read_index + code > length && code != 1)
            return 0; // Malformed
        read_index++;

        for (i = 1; i < code; i++)
        {
            output[write_index++] = input[read_index++];
        }

        if (code != 0xFF && read_index != length)
        {
            output[write_index++] = 0x00;
        }
    }
    return write_index;
}

static size_t cobs_encode(const uint8_t *input, size_t length, uint8_t *output)
{
    size_t read_index = 0;
    size_t write_index = 1; // Start after the first code byte
    uint8_t code_index = 0; // Index of the current code byte
    uint8_t code = 1;       // Code starts at 1

    while (read_index < length)
    {
        if (input[read_index] == 0)
        {
            output[code_index] = code;
            code_index = write_index++;
            code = 1;
            read_index++;
        }
        else
        {
            output[write_index++] = input[read_index++];
            code++;
            if (code == 0xFF)
            {
                output[code_index] = code;
                code_index = write_index++;
                code = 1;
            }
        }
    }
    output[code_index] = code; // Write final code byte
    output[write_index++] = FRAME_DELIMITER; // Frame delimiter
    return write_index;
}

// Parses the decoded byte buffer
bool parse_packet(const uint8_t *encoded_buffer, size_t encoded_len, Packet *pkt)
{
    uint8_t decoded[64]; // Ensure this is large enough for CMD + LEN + PAYLOAD + CHK
    uint32_t decoded_len = cobs_decode(encoded_buffer, encoded_len, decoded);

    if (decoded_len < 3)
        return false; // Minimum size: CMD, LEN, CHK

    pkt->cmd = decoded[0];
    pkt->len = decoded[1];

    if (decoded_len != (uint32_t)(3 + pkt->len))
        return false; // Length mismatch

    uint8_t calc_chk = pkt->cmd ^ pkt->len;
    for (uint8_t i = 0; i < pkt->len; i++)
    {
        pkt->payload[i] = decoded[2 + i];
        calc_chk ^= pkt->payload[i];
    }

    pkt->checksum = decoded[2 + pkt->len];
    return pkt->checksum == calc_chk;
}

size_t create_packet(uint8_t cmd, const uint8_t *payload, size_t payload_len, uint8_t *output)
{
    if (payload_len > MAX_PAYLOAD)
        return 0; // Payload too large

    uint8_t buffer[3 + MAX_PAYLOAD]; // CMD + LEN + PAYLOAD + CHK
    buffer[0] = cmd;
    buffer[1] = payload_len;

    uint8_t checksum = cmd ^ payload_len;
    for (size_t i = 0; i < payload_len; i++)
    {
        buffer[2 + i] = payload[i];
        checksum ^= payload[i];
    }
    buffer[2 + payload_len] = checksum;

    return cobs_encode(buffer, 3 + payload_len, output);
}
